### PR TITLE
Pro 3956 stop writing all postcode addresses to form data

### PR DESCRIPTION
**Pull request by harjeetkalsi**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/PRO-3956

### CODE REVIEW

File: app/core/steps/AddressStep.js [hunk 1]
```diff
@@ -16,15 +16,22 @@ class AddressStep extends ValidationStep {
         return [ctx];
     }
 
-    handlePost(ctx, errors) {
+    handlePost(ctx, errors, formdata, session) {
         ctx.address = ctx.postcodeAddress || ctx.freeTextAddress;
         ctx.postcode = ctx.postcode ? ctx.postcode.toUpperCase() : ctx.postcode;
-        if (!ctx.postcodeAddress) {
+        if (!ctx.postcodeAddress){
```
Comment (tomxelliott) [inline app/core/steps/AddressStep.js]: Is the lack of space not breaking esLint?

File: app/steps/ui/applicant/address/index.js [hunk 2]
```diff
@@ -7,6 +7,22 @@ class ApplicantAddress extends AddressStep {
     static getUrl() {
         return '/applicant-address';
     }
+    getContextData(req){
+        const ctx =super.getContextData(req);
```
Comment (tomxelliott) [inline app/steps/ui/applicant/address/index.js]: the lack of space between `=` and `super` must be breaking esLint?

File: app/steps/ui/deceased/address/index.js [hunk 3]
```diff
@@ -9,10 +9,21 @@ class DeceasedAddress extends AddressStep {
         return '/deceased-address';
     }
 
+    getContextData(req){
+        const ctx =super.getContextData(req);
```
Comment (tomxelliott) [inline app/steps/ui/deceased/address/index.js]: the lack of space between `=` and `super` must be breaking esLint?

File: test/unit/testApplicantAddress.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        it('should return the correct url', (done) => {
+            const url = ApplicantAddress.constructor.getUrl();
+            expect(url).to.equal('/applicant-address');
+            done();
+        });
+    });
+
+    describe('getContextData()', () => {
+        let ctx;
+        it('should return the ctx with the deceased addresses', (done) => {
```
Comment (shepda) [inline test/unit/testApplicantAddress.js]: This test is testing the applicant not the deceased so can you update the test description

File: app/steps/ui/executors/address/index.js [hunk 5]
```diff
@@ -22,6 +22,9 @@ class ExecutorAddress extends AddressStep {
         } else if (startsWith(req.path, path)) {
             ctx.index = this.recalcIndex(ctx, 0);
         }
+        if (req.session.addresses && req.session.addresses.executors && req.session.addresses.executors[ctx.index-1]) {
+            ctx.addresses = req.session.addresses.executors[ctx.index-1];
+        }
```
Comment (shepda) [inline app/steps/ui/executors/address/index.js]: Can we have unit tests to test this logic?

File: app/steps/ui/executors/address/index.js [hunk 6]
```diff
@@ -32,9 +35,8 @@ class ExecutorAddress extends AddressStep {
         if (ctx.list[ctx.index].address) {
             ctx.address = ctx.list[ctx.index].postcodeAddress || ctx.list[ctx.index].freeTextAddress;
             ctx.postcode = ctx.list[ctx.index].postcode;
-            if (ctx.list[ctx.index].postcodeAddress) {
-                ctx.addresses = [{formatted_address: ctx.address}];
-            } else {
+            ctx.postcodeAddress = ctx.list[ctx.index].postcodeAddress;
+            if (!ctx.list[ctx.index].postcodeAddress) {
```
Comment (shepda) [inline app/steps/ui/executors/address/index.js]: Can we have unit tests to test this logic?

File: app/steps/ui/executors/address/index.js [hunk 7]
```diff
@@ -54,6 +56,12 @@ class ExecutorAddress extends AddressStep {
         if (ctx.index === -1) {
             ctx.allExecsApplying = ctx.executorsWrapper.areAllAliveExecutorsApplying();
         }
+        if (ctx.addresses && !session.addresses.executors) {
+            session.addresses.executors = [];
+        }
+        if (ctx.addresses) {
+            session.addresses.executors.push(ctx.addresses);
+        }
```
Comment (shepda) [inline app/steps/ui/executors/address/index.js]: Can we have unit tests to test this logic?

File: app/core/steps/AddressStep.js [hunk 8]
```diff
@@ -16,15 +16,22 @@ class AddressStep extends ValidationStep {
...
     }
 
-    handlePost(ctx, errors) {
+    handlePost(ctx, errors, formdata, session) {
         ctx.address = ctx.postcodeAddress || ctx.freeTextAddress;
         ctx.postcode = ctx.postcode ? ctx.postcode.toUpperCase() : ctx.postcode;
         if (!ctx.postcodeAddress) {
             delete ctx.addresses;
         }
+        session.addresses = session.addresses || {};
```
Comment (shepda) [inline app/core/steps/AddressStep.js]: Can we have a unit test to test that if `session.addresses` doesn't exist then it gets set to `{}`?

Comment (harjeetkalsi) [reply app/core/steps/AddressStep.js]: yes, i think is covered in testAddressStep.js in "it('when freeTextAddress exists',(done..." where it is checked there is a session.addresses{} after the handle post.

**Comment by harjeetkalsi** [comment]
@shepda in reply to yarn eslint, the 1 error it gives me is that a 'ctx' is assigned but never used, on 166 on testApplicantAddress.js , however if I delete the 'ctx' the test fails.

File: test/unit/testApplicantAddress.js (hunk 4 above)
Comment (harjeetkalsi) [reply test/unit/testApplicantAddress.js]: done!

File: app/steps/ui/executors/address/index.js (hunk 5 above)
Comment (harjeetkalsi) [reply app/steps/ui/executors/address/index.js]: yep, have now added new test in testExecutorAddress.js called 'should return the ctx with the executors addresses for the correct index'

File: app/core/steps/AddressStep.js (hunk 8 above)
Comment (shepda) [reply app/core/steps/AddressStep.js]: 👍

File: app/steps/ui/executors/address/index.js (hunk 7 above)
Comment (harjeetkalsi) [reply app/steps/ui/executors/address/index.js]: ah yes that is the 'returns the correct data and errors' in the handlePost() in testExecutorAddress

**Pull request closed without merging**